File: rv-error/src/main/java/com/runtimeverification/error/CSVRenderer.java
// Copyright (c) 2017 Runtime Verification, Inc. (RV-Match team). All Rights Reserved.
package com.runtimeverification.error;

import com.runtimeverification.error.data.Citation;
import com.runtimeverification.error.data.ErrorCategory;
import com.runtimeverification.error.data.LocationError;
import com.runtimeverification.error.data.Metadata;
import com.runtimeverification.error.data.StackError;
import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;

import java.io.IOException;
import java.io.PrintStream;
import java.util.List;

/**
 * Created by dwightguth on 1/4/17.
 */
public class CSVRenderer extends Renderer {
    public CSVRenderer(Metadata data) {
        super(data);
    }

    @Override
    protected void renderImpl(StackError error, PrintStream out) {
        CSVPrinter printer = getCSVPrinter(out);
        String[] strings = getCommonFields(error.error_id, error.description, error.category, error.citations);

        StringBuilder reason = new StringBuilder();
        error.traces.subList(1, error.traces.size()).forEach(t -> getConsole().renderTrace(reason, t));
        strings[4] = reason.toString();

        StringBuilder trace = new StringBuilder();
        if (error.traces.size() > 0) {
            getConsole().renderTrace(trace, error.traces.get(0));
        }
        strings[5] = trace.toString();

        printRecord(printer, strings);
    }

    private void printRecord(CSVPrinter printer, String[] strings) {
        try {
            printer.printRecord(strings);
            printer.flush();
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(2);
            throw new AssertionError("unreachable");
        }
    }

    private String[] getCommonFields(String error_id, String description, ErrorCategory category, List<Citation> citations) {
        String[] strings = new String[6];
        strings[0] = error_id;
        strings[1] = description;
        strings[2] = renderErrorCategory(category);

        StringBuilder citationsSb = new StringBuilder();
        citations.forEach(c -> renderCitation(citationsSb, c));
        strings[3] = citationsSb.toString();
        return strings;
    }

    private void renderCitation(StringBuilder sb, Citation citation) {
        sb.append("\n");
        sb.append(citation.document);
        sb.append(",");
        sb.append(citation.section);
        sb.append(",");
        if (citation.paragraph != null) {
            sb.append(citation.paragraph);
        }
        sb.append(",");
        getConsole().renderCitationUrl(sb, citation);
    }

    private CSVPrinter getCSVPrinter(PrintStream out) {
        try {
            return new CSVPrinter(out, CSVFormat.EXCEL);
        } catch (IOException e) {
            e.printStackTrace();
            System.exit(2);
            throw new AssertionError("unreachable");
        }
    }

    @Override
    protected void renderImpl(LocationError error, PrintStream out) {
        CSVPrinter printer = getCSVPrinter(out);
        String[] strings = getCommonFields(error.error_id, error.description, error.category, error.citations);
        strings[4] = "";

        StringBuilder loc = new StringBuilder();
        getConsole().renderLoc(loc, error.loc.loc);
        strings[5] = loc.toString();

        printRecord(printer, strings);
    }


}

File: errors/rv-error/src/main/ocaml/csv_renderer.ml
open Base_renderer
open Error_t
open Rv_error_util


let default (d : 'a) (x : 'a option) : 'a =
  match x with
  | None -> d
  | Some x -> x


let render_citation (buffer : rv_buffer) (citation : citation) : unit =
  add_line buffer "" ;
  concat buffer ',' [
    citation.document;
    citation.section;
    default "" citation.paragraph;
    render_citation_url citation;
  ]


let get_common_fields (error_id : string) (description : string) (category : error_category) (citations : citation list) : string list =
  let cat = string_of_error_category category in
  let cit_buffer = create_buffer 256 in
  List.iter (render_citation cit_buffer) citations ;
  [error_id; description; cat; (Buffer.contents cit_buffer.data)]


let print_record (printer : out_channel) (strings : string list) : unit =
  try
    Csv.output_record (Csv.to_channel ~excel_tricks:true printer) strings ;
    flush printer
  with e ->
    Rv_error_util.print_stack_trace e ;
    exit 2


let render_stack_error_impl (this : renderer) (error : stack_error) (str : out_channel) =
  let common_fields = get_common_fields error.error_id error.description error.category error.citations in
  let uncommon_fields =
    match error.stack_traces with
    | [] -> [""; ""]
    | frame::traces ->
      let buf_frame = create_buffer 256 in
      render_trace buf_frame frame ;
      let buf_traces = create_buffer 256 in
      List.iter (render_trace buf_traces) traces ;
      [Buffer.contents buf_frame.data; Buffer.contents buf_traces.data]
  in
  let strings = List.append common_fields uncommon_fields in
  print_record str strings


let render_location_error_impl (this : renderer) (error : location_error) (str : out_channel) =
  let common_fields = get_common_fields error.error_id error.description error.category error.citations in
  let buf_loc = create_buffer 256 in
  render_loc buf_loc error.loc.loc ;
  let strings = List.append common_fields [""; Buffer.contents buf_loc.data] in
  print_record str strings

let render_impl (this : renderer) (error : rv_error) (str : out_channel) : unit =
  match error with
  | StackError error -> render_stack_error_impl this error str
  | LocationError error -> render_location_error_impl this error str


let instance (renderer : renderer) : renderer =
  Base_renderer.create_instance renderer render_impl
